internal/querycoordv2/meta/replica_manager.go

```diff
@@ -0,0 +1,191 @@
+	if err != nil {
+		return fmt.Errorf("failed to recover replicas, err=%w", err)
+	}
+	for _, replica := range replicas {
+		m.replicas[replica.GetID()] = &Replica{
+			Replica: replica,
+			Nodes:   NewUniqueSet(replica.GetNodes()...),
+		}
+	}
+	return nil
+}
+
+func (m *ReplicaManager) Get(id UniqueID) *Replica {
+	m.rwmutex.RLock()
+	defer m.rwmutex.RUnlock()
+
+	return m.replicas[id]
+}
+
+// Spawn spawns replicas of the given number, for given collection,
+// this doesn't store these replicas and assign nodes to them.
+func (m *ReplicaManager) Spawn(collection int64, replicaNumber int32) ([]*Replica, error) {
+	var (
+		replicas = make([]*Replica, replicaNumber)
+		err      error
+	)
+	for i := range replicas {
+		replicas[i], err = m.spawn(collection)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return replicas, err
+}
+
+func (m *ReplicaManager) Put(replicas ...*Replica) error {
+	m.rwmutex.Lock()
+	defer m.rwmutex.Unlock()
+
+	return m.put(replicas...)
+}
+
+func (m *ReplicaManager) spawn(collectionID UniqueID) (*Replica, error) {
+	id, err := m.idAllocator()
+	if err != nil {
+		return nil, err
+	}
+	return &Replica{
+		Replica: &querypb.Replica{
+			ID:           id,
+			CollectionID: collectionID,
+		},
+		Nodes: make(UniqueSet),
+	}, nil
+}
+
+func (m *ReplicaManager) put(replicas ...*Replica) error {
+	for _, replica := range replicas {
+		err := m.store.SaveReplica(replica.Replica)
+		if err != nil {
+			return err
+		}
+		m.replicas[replica.ID] = replica
+	}
+	return nil
+}
+
+// RemoveCollection removes replicas of given collection,
+// returns error if failed to remove replica from KV
+func (m *ReplicaManager) RemoveCollection(collectionID UniqueID) error {
+	m.rwmutex.Lock()
+	defer m.rwmutex.Unlock()
+
+	err := m.store.ReleaseReplicas(collectionID)
+	if err != nil {
+		return err
+	}
+	for id, replica := range m.replicas {
+		if replica.CollectionID == collectionID {
+			delete(m.replicas, id)
+		}
+	}
+	return nil
+}
+
+func (m *ReplicaManager) GetByCollection(collectionID UniqueID) []*Replica {
+	m.rwmutex.RLock()
+	defer m.rwmutex.RUnlock()
+
+	replicas := make([]*Replica, 0, 3)
+	for _, replica := range m.replicas {
+		if replica.CollectionID == collectionID {
+			replicas = append(replicas, replica)
+		}
+	}
+
+	return replicas
+}
+
+func (m *ReplicaManager) GetByCollectionAndNode(collectionID, nodeID UniqueID) *Replica {
+	m.rwmutex.RLock()
+	defer m.rwmutex.RUnlock()
+
+	for _, replica := range m.replicas {
+		if replica.CollectionID == collectionID && replica.Nodes.Contain(nodeID) {
+			return replica
+		}
+	}
+
+	return nil
+}
+
+func (m *ReplicaManager) AddNode(replicaID UniqueID, nodes ...UniqueID) error {
+	m.rwmutex.Lock()
+	defer m.rwmutex.Unlock()
+
+	replica, ok := m.replicas[replicaID]
+	if !ok {
+		return ErrReplicaNotFound
+	}
+
+	replica = replica.Clone()
+	replica.AddNode(nodes...)
+	return m.put(replica)
+}
+
+func (m *ReplicaManager) RemoveNode(replicaID UniqueID, nodes ...UniqueID) error {
```

Comment:
Repica manager need a way to rebalance.

For example, by default there are 3 nodes.
Replica1. -> 2 nodes Replica2 -> 1 nodes.
if replica 2 node down, then we have to rebalance between replica1 and replica2.
The rebalance can be done in a another checker